jruby 9.3.x: stacktrace is extended on each raise

This issue seems to be introduced in jruby 9.3.0.0 and still exists in latest 9.3.4.0.
9.2.20.1 seems not to be affected.

If you call a script multiple times that raises an error the stacktrace gets bigger and bigger.
for example call a script `raise 'foo'` in java, catch the error and then call another script `raise 'bar'` (or the same as before).
In jruby 9.2.x the second stacktrace looks like this
```
...
Caused by: org.jruby.embed.EvalFailedException: (RuntimeError) bar
	at org.jruby.embed.internal.EmbedEvalUnitImpl.run(EmbedEvalUnitImpl.java:131)
	at org.jruby.embed.jsr223.JRubyEngine.eval(JRubyEngine.java:90)
	... 26 more
Caused by: org.jruby.exceptions.RuntimeError: (RuntimeError) bar
	at RUBY.<main>(<script>:1)
```

in jruby 9.3.x the second stacktrace looks like this
```
...
Caused by: org.jruby.exceptions.RuntimeError: (RuntimeError) bar
	at RUBY.<main>(<script>:1)
Caused by: org.jruby.exceptions.RuntimeError: (RuntimeError) foo
	... 1 more
```

Here is a simple test-case
```
  @Test
  public void testRubyRaise()
    throws Exception
  {
    try
    {
      final ScriptEngine _jruby1 = new JRubyEngineFactory().getScriptEngine();
      _jruby1.eval("raise 'foo'");
    }
    catch (Exception _e)
    {
      _e.printStackTrace(System.err);
      try (ByteArrayOutputStream _bos = new ByteArrayOutputStream();
          PrintStream _s = new PrintStream(_bos))
      {
        _e.printStackTrace(_s);
        Assert.assertFalse(_bos.toString().contains("bar"), "stack contains bar: " + _bos.toString());
      }
    }
    try
    {
      final ScriptEngine _jruby2 = new JRubyEngineFactory().getScriptEngine();
      _jruby2.eval("raise 'bar'");
    }
    catch (Exception _e)
    {
      _e.printStackTrace(System.err);
      try (ByteArrayOutputStream _bos = new ByteArrayOutputStream();
          PrintStream _s = new PrintStream(_bos))
      {
        _e.printStackTrace(_s);
        Assert.assertFalse(_bos.toString().contains("foo"), "stack contains foo: " + _bos.toString());
      }
    }
  }
```

--- a/core/src/main/java/org/jruby/embed/internal/EmbedEvalUnitImpl.java
+++ b/core/src/main/java/org/jruby/embed/internal/EmbedEvalUnitImpl.java
@@ -124,6 +124,9 @@ public class EmbedEvalUnitImpl implements EmbedEvalUnit {
             return ret;
         }
         catch (RaiseException e) {
+            // Exception is about to be propagated out to Java, so clear $!
+            runtime.getCurrentContext().setErrorInfo(runtime.getNil());
+
             // handle exits as simple script termination
             if ( e.getException() instanceof RubySystemExit ) {
                 return ((RubySystemExit) e.getException()).status();

--- a/core/src/main/java/org/jruby/embed/jsr223/JRubyEngine.java
+++ b/core/src/main/java/org/jruby/embed/jsr223/JRubyEngine.java
@@ -42,6 +42,8 @@ import javax.script.ScriptEngineFactory;
 import javax.script.ScriptException;
 import javax.script.SimpleBindings;
 import javax.script.SimpleScriptContext;
+
+import org.jruby.Ruby;
 import org.jruby.embed.EmbedEvalUnit;
 import org.jruby.embed.ScriptingContainer;
 import org.jruby.exceptions.NoMethodError;
@@ -274,6 +276,11 @@ public class JRubyEngine implements Compilable, Invocable, ScriptEngine {
 
             ScriptException se = new ScriptException("Error during evaluation of Ruby in " + file + " at line " + line + ": " + e.getMessage());
             se.initCause(e);
+
+            // Exception is about to be propagated out to Java, so clear $!
+            Ruby runtime = e.getException().getRuntime();
+            runtime.getCurrentContext().setErrorInfo(runtime.getNil());
+
             return se;
         }
 

--- a/core/src/test/java/org/jruby/embed/jsr223/JRubyEngineTest.java
+++ b/core/src/test/java/org/jruby/embed/jsr223/JRubyEngineTest.java
@@ -827,6 +827,38 @@ public class JRubyEngineTest extends BaseTest {
         assertNotNull( getVarMap(instance).getVariable("ARGV") );
     }
 
+    @Test
+    public void testRaiseDoesNotStackCause()
+            throws Exception
+    {
+        try
+        {
+            final ScriptEngine _jruby1 = new JRubyEngineFactory().getScriptEngine();
+            _jruby1.eval("raise 'foo'");
+        }
+        catch (Exception _e)
+        {
+            try (ByteArrayOutputStream _bos = new ByteArrayOutputStream();
+                 PrintStream _s = new PrintStream(_bos))
+            {
+                assertFalse("stack contains bar: " + _bos.toString(), _bos.toString().contains("bar"));
+            }
+        }
+        try
+        {
+            final ScriptEngine _jruby2 = new JRubyEngineFactory().getScriptEngine();
+            _jruby2.eval("raise 'bar'");
+        }
+        catch (Exception _e)
+        {
+            try (ByteArrayOutputStream _bos = new ByteArrayOutputStream();
+                 PrintStream _s = new PrintStream(_bos))
+            {
+                assertFalse("stack contains foo: " + _bos.toString(), _bos.toString().contains("foo"));
+            }
+        }
+    }
+
     private static BiVariableMap getVarMap(final ScriptEngine engine) {
         return ((JRubyEngine) engine).container.getVarMap();
     }

